File: backend/internal/model/memory.go
package model

import "time"

type Memory struct {
	ID        string    `json:"id"`
	UserID    string    `json:"user_id"`
	Type      string    `json:"type"`
	Content   string    `json:"content"`
	Summary   string    `json:"summary"`
	Title     string    `json:"title"`
	Tags      []string  `json:"tags"`
	MediaURL  string    `json:"media_url,omitempty"`
	CreatedAt time.Time `json:"created_at"`
	UpdatedAt time.Time `json:"updated_at"`
}

func (m Memory) ToJSON() map[string]interface{} {
	return map[string]interface{}{
		"id":          m.ID,
		"user_id":     m.UserID,
		"type":        m.Type,
		"content":     m.Content,
		"summary":     m.Summary,
		"title":       m.Title,
		"tags":        m.Tags,
		"media_url":   m.MediaURL,
		"created_at":  m.CreatedAt,
		"updated_at":  m.UpdatedAt,
	}
}

func JSONToMemory(m map[string]interface{}) Memory {
	tags := []string{}
	if t, ok := m["tags"].([]string); ok {
		tags = t
	}
	return Memory{
		ID:       getString(m, "id"),
		UserID:   getString(m, "user_id"),
		Type:     getString(m, "type"),
		Content:  getString(m, "content"),
		Summary:  getString(m, "summary"),
		Title:    getString(m, "title"),
		Tags:     tags,
		MediaURL: getString(m, "media_url"),
	}
}

func getString(m map[string]interface{}, key string) string {
	if v, ok := m[key].(string); ok {
		return v
	}
	return ""
}

type User struct {
	ID        string    `json:"id"`
	Phone     string    `json:"phone"`
	Email     string    `json:"email"`
	CreatedAt time.Time `json:"created_at"`
}

func (u User) ToJSON() map[string]interface{} {
	return map[string]interface{}{
		"id":    u.ID,
		"phone": u.Phone,
		"email": u.Email,
	}
}

type ChatMessage struct {
	Message string `json:"message"`
}

type ChatReply struct {
	Reply string `json:"reply"`
}

File: lib/models/memory.dart
class Memory {
  final String id;
  final String userId;
  final String type; // "voice", "text", "image"
  final String content;
  final String summary;
  final String title;
  final List<String> tags;
  final String? mediaUrl;
  final DateTime createdAt;
  final DateTime updatedAt;

  Memory({
    required this.id,
    required this.userId,
    required this.type,
    required this.content,
    this.summary = '',
    this.title = '',
    this.tags = const [],
    this.mediaUrl,
    required this.createdAt,
    required this.updatedAt,
  });

  factory Memory.fromJson(Map<String, dynamic> json) {
    return Memory(
      id: json['id'] ?? '',
      userId: json['user_id'] ?? '',
      type: json['type'] ?? 'text',
      content: json['content'] ?? '',
      summary: json['summary'] ?? '',
      title: json['title'] ?? '',
      tags: List<String>.from(json['tags'] ?? []),
      mediaUrl: json['media_url'],
      createdAt: DateTime.tryParse(json['created_at'] ?? '') ?? DateTime.now(),
      updatedAt: DateTime.tryParse(json['updated_at'] ?? '') ?? DateTime.now(),
    );
  }

  Map<String, dynamic> toJson() => {
    'id': id,
    'user_id': userId,
    'type': type,
    'content': content,
    'summary': summary,
    'title': title,
    'tags': tags,
    'media_url': mediaUrl,
    'created_at': createdAt.toIso8601String(),
    'updated_at': updatedAt.toIso8601String(),
  };

  Memory copyWith({
    String? id, String? userId, String? type, String? content,
    String? summary, String? title, List<String>? tags, String? mediaUrl,
    DateTime? createdAt, DateTime? updatedAt,
  }) {
    return Memory(
      id: id ?? this.id,
      userId: userId ?? this.userId,
      type: type ?? this.type,
      content: content ?? this.content,
      summary: summary ?? this.summary,
      title: title ?? this.title,
      tags: tags ?? this.tags,
      mediaUrl: mediaUrl ?? this.mediaUrl,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
    );
  }
}
